src/cli/cli.cc

```diff
@@ -157,9 +160,59 @@ static CheckInResult checkIn(AkliteClientExt &client, CheckMode check_mode,
   }
 }
 
-StatusCode CheckIn(AkliteClientExt &client, const LocalUpdateSource *local_update_source, CheckMode check_mode) {
+static Json::Value getCheckInResultJson(AkliteClientExt &client, const CheckInResult &ci_res,
+                                        const GetTargetToInstallResult &gti_res) {
+  const auto current = client.GetCurrent();
+
+  Json::Value json_root;
+
+  auto app_shortlist = client.GetAppShortlist();
+  for (const auto &target : ci_res.Targets()) {
+    Json::Value json_target;
+    json_target["name"] = target.Name();
+    json_target["version"] = target.Version();
+    if (!gti_res.selected_target.IsUnknown() && gti_res.selected_target.Name() == target.Name()) {
+      json_target["selected"] = true;
+      json_target["reason"] = gti_res.reason;
+    }
+    if (client.IsRollback(target)) {
```

Comment:
This is not quite efficient performance wise since each `.IsRollback()` call does exactly the same action - get installed versions from the DB. I wonder if it is possible to optimize it somehow since a user device can have thousands targets?

Reply:
We could add a method to `LiteClient` returning a `set` with the names of all targets that caused a rollback. It would be just a  matter of using functionality from `get_known_but_not_installed_versions` and a little bit from `known_local_target`. What do you think?

Reply:
I thought about avoiding this iteration over targets here at all. Instead, we can extend the `filterTargets` functionality in the `AkliteClient::CheckIn()` to add the "bad" mark (another attribute of TufTarget object, or additional field in the custom json) (1).

The other two marks "current" and "selected" seems reasonable to set directly in the json output `getCheckInResultJson`. However, if we get more than one "client" for it, we may wanna add another two attributes to TufTarget in `AkliteClientExt::GetTargetToInstall`, but it means the input checkin result modification...(2)

I think it is more or less reasonable to do the first item now, unless it requires too much effort.

Reply:
I'm not sure about that, will give it a little more thought. Adding an attribute to `TufTarget`, or event to the custom json, would create an expectation on that attribute to always be correctly set, which has its price, while not being used that much.

Reply:
OK, let's not think about performance optimization at this stage.

Reply:
OK. This is something we can handle in a separate PR, as the potential performance hit is only related to the new feature.